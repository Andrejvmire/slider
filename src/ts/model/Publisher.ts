import {isArray, isNil, isUndefined} from "lodash";

class Publisher implements IPublisher {
    private observer: Map<TPublisher, Set<ISubscriber>>;

    constructor() {
        this.init();
    }

    init() {
        this.observer = new Map<TPublisher, Set<ISubscriber>>();
    }

    attach(subscriber: ISubscriber, type: TPublisher): void {
        let subscribers = this.observer.get(type);
        if (isUndefined(subscribers)) {
            subscribers = new Set<ISubscriber>()
        }
        this.observer.set(type, subscribers.add(subscriber));
    }

    detach(subscriber: ISubscriber, type: TPublisher): void {
        let keys: IterableIterator<TPublisher>;
        if (isNil(type)) {
            keys = this.observer.keys();
        } else {
            keys = [type][Symbol.iterator]();
        }
        // eslint-disable-next-line no-constant-condition
        while (true) {
            const result = keys.next();
            if (result.done === true) break;
            if (this.observer.has(result.value)) {
                this.observer.get(result.value)?.delete(subscriber);
            }
        }
    }

    notify(type: TPublisher): void;
    notify(type: TPublisher[]): void;
    notify(type: TPublisher | TPublisher[]): void {
        if (isArray(type)) {
            type.forEach(this.notify.bind(this));
        } else {
            this.observer
                .get(type)
                ?.forEach(
                    subscriber => subscriber.update(type)
                );
        }
    }

}

export default Publisher;