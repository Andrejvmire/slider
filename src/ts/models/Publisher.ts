abstract class Publisher implements IPublisher {
    protected abstract _subscribers: ISubscriber[] = [];

    abstract values(): any;

    attach(subscriber: ISubscriber): void {
        const isExist = this._subscribers.includes(subscriber);
        if (!isExist) {
            this._subscribers.push(subscriber);
        }
    }

    detach(subscriber: ISubscriber): void {
        const index = this._subscribers.indexOf(subscriber);
        if (index !== -1) {
            this._subscribers.splice(index, 1);
        }
    }

    notify(): void {
        for (const subscriber of this._subscribers) {
            subscriber.update(this.values())
        }
    }
}

export default Publisher;