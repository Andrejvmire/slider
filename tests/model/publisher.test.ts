import "reflect-metadata";
import Publisher from "../../src/ts/model/Publisher";
import {container} from "tsyringe";

describe("Тест издателя", function () {
    const subscriber_1 = {
            update: jest.fn((publisher: TPublisher) => publisher)
        },
        subscriber_2 = {
            update: jest.fn((publisher: TPublisher) => publisher)
        },
        error: TPublisher = "error",
        type: TPublisher = "apply";
    let publisher: IPublisher;
    it("Создаем без ошибки", () => {
        expect(() => {
            container.register("IPublisher", Publisher);
            publisher = container.resolve<IPublisher>("IPublisher");
        }).not.toThrowError();
    })
    it('Отписываем несуществующего подписчика', () => {
        expect(() => publisher.detach(subscriber_1, error)).not.toThrowError();
    });
    it("Выполнение подписки на издателя", () => {
        publisher.attach(subscriber_1, type);
        expect(publisher).toMatchSnapshot();
    });
    it('Оповещает подписчика', () => {
        publisher.notify(type);
        expect(subscriber_1.update.mock.calls).toEqual([["apply"]]);
    });
    it('Пытаемся подписаться еще раз', () => {
        publisher.attach(subscriber_1, type);
        expect(publisher).toMatchSnapshot();
    });
    it('Добавляем еще одного (другого) подписчика', () => {
        publisher.attach(subscriber_2, type);
        expect(publisher).toMatchSnapshot();
    });
    it('Оповещаем несколько групп', () => {
        expect(() => publisher.notify(["error", "apply", "change"]))
            .not.toThrowError();
    });
    it('Выполнит отписку 1-го', () => {
        publisher.detach(subscriber_1, type);
        expect(publisher).toMatchSnapshot();
    });
    it('Выполнит отписку 2-го', () => {
        publisher.detach(subscriber_2, type);
        expect(publisher).toMatchSnapshot();
    });
    it("Подписываем сразу на 2 типа рассылки", () => {
        publisher.attach(subscriber_1, [type, error]);
        expect(publisher).toMatchSnapshot();
    })
})