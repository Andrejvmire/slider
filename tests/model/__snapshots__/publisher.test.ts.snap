// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Тест издателя Выполнение подписки на издателя 1`] = `
Publisher {
  "observer": Map {
    "apply" => Set {
      Object {
        "update": [MockFunction],
      },
    },
  },
}
`;

exports[`Тест издателя Выполнит отписку 1-го 1`] = `
Publisher {
  "observer": Map {
    "apply" => Set {
      Object {
        "update": [MockFunction] {
          "calls": Array [
            Array [
              "apply",
            ],
          ],
          "results": Array [
            Object {
              "type": "return",
              "value": "apply",
            },
          ],
        },
      },
    },
  },
}
`;

exports[`Тест издателя Выполнит отписку 2-го 1`] = `
Publisher {
  "observer": Map {
    "apply" => Set {},
  },
}
`;

exports[`Тест издателя Добавляем еще одного (другого) подписчика 1`] = `
Publisher {
  "observer": Map {
    "apply" => Set {
      Object {
        "update": [MockFunction] {
          "calls": Array [
            Array [
              "apply",
            ],
          ],
          "results": Array [
            Object {
              "type": "return",
              "value": "apply",
            },
          ],
        },
      },
      Object {
        "update": [MockFunction],
      },
    },
  },
}
`;

exports[`Тест издателя Подписываем сразу на 2 типа рассылки 1`] = `
Publisher {
  "observer": Map {
    "apply" => Set {
      Object {
        "update": [MockFunction] {
          "calls": Array [
            Array [
              "apply",
            ],
            Array [
              "apply",
            ],
          ],
          "results": Array [
            Object {
              "type": "return",
              "value": "apply",
            },
            Object {
              "type": "return",
              "value": "apply",
            },
          ],
        },
      },
    },
    "error" => Set {
      Object {
        "update": [MockFunction] {
          "calls": Array [
            Array [
              "apply",
            ],
            Array [
              "apply",
            ],
          ],
          "results": Array [
            Object {
              "type": "return",
              "value": "apply",
            },
            Object {
              "type": "return",
              "value": "apply",
            },
          ],
        },
      },
    },
  },
}
`;

exports[`Тест издателя Пытаемся подписаться еще раз 1`] = `
Publisher {
  "observer": Map {
    "apply" => Set {
      Object {
        "update": [MockFunction] {
          "calls": Array [
            Array [
              "apply",
            ],
          ],
          "results": Array [
            Object {
              "type": "return",
              "value": "apply",
            },
          ],
        },
      },
    },
  },
}
`;
